packages/playwright-core/src/server/trace/recorder/tracing.ts

```diff
@@ -570,18 +528,75 @@ function createAfterActionTraceEvent(metadata: CallMetadata): trace.AfterActionT
   };
 }
 
-async function appendEventAndFlushIfNeeded(file: { file: string, buffer: trace.TraceEvent[] }, event: trace.TraceEvent) {
-  file.buffer.push(event);
+class SerializedFS {
+  private _writeChain = Promise.resolve();
+  private _buffers = new Map<string, string[]>();
 
-  // Do not flush events, they are too noisy.
-  if (event.type === 'event' || event.type === 'object')
-    return;
+  mkdir(dir: string) {
+    this._appendOperation(() => fs.promises.mkdir(dir, { recursive: true }));
+  }
 
-  await flushTraceFile(file);
-}
+  writeFile(file: string, content: string | Buffer, skipIfExists?: boolean) {
+    // Note: 'wx' flag only writes when the file does not exist.
+    // See https://nodejs.org/api/fs.html#file-system-flags.
+    // This way tracing never have to write the same resource twice.
+    this._appendOperation(async () => {
+      if (skipIfExists)
+        await fs.promises.writeFile(file, content, { flag: 'wx' }).catch(() => {});
+      else
+        await fs.promises.writeFile(file, content);
+    });
+  }
+
+  appendFile(file: string, text: string, flush?: boolean) {
+    if (!this._buffers.has(file))
+      this._buffers.set(file, []);
+    this._buffers.get(file)!.push(text);
+    if (flush)
+      this.flushFile(file);
+  }
+
+  flushFile(file: string) {
+    const buffer = this._buffers.get(file) || [];
+    const text = buffer.join('');
+    this._buffers.delete(file);
+    this._appendOperation(() => fs.promises.appendFile(file, text));
+  }
 
-async function flushTraceFile(file: { file: string, buffer: trace.TraceEvent[] }) {
-  const data = file.buffer.map(e => Buffer.from(JSON.stringify(e) + '\n'));
-  await fs.promises.appendFile(file.file, Buffer.concat(data));
-  file.buffer = [];
+  copyFile(from: string, to: string) {
+    this._appendOperation(() => fs.promises.copyFile(from, to));
+  }
+
+  async flushEverything() {
+    for (const file of this._buffers.keys())
+      this.flushFile(file);
+    await this._writeChain;
+  }
+
+  exportZip(entries: NameValue[], zipFileName: string, onError: (error: Error) => void) {
+    // Chain the export operation against write operations,
+    // so that files do not change during the export.
+    this._appendOperation(async () => {
+      const zipFile = new yazl.ZipFile();
+      const result = new ManualPromise<Error | void>();
+      (zipFile as any as EventEmitter).on('error', error => result.resolve(error));
```

Comment:
```suggestion
      (zipFile as any as EventEmitter).on('error', () => result.resolve());
```

I feel like this is not used here anymore.